crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs

```diff
@@ -362,111 +348,67 @@ impl RetriableRpcClient {
         sequence_number: u64,
     ) -> Result<CheckpointData, RetriableClientError> {
         let _scope = monitored_scope::monitored_scope("RetriableRpcClient::get_full_checkpoint");
-        let start_time = Instant::now();
-
-        // Check if we should directly skip RPC node due to previous failures, and use fallback.
-        // When the fallback client is not configured, we should always use the RPC node.
-        let try_rpc_node_first = self.fallback_client.is_none()
-            || start_time >= self.skip_rpc_for_checkpoint_until.load(Ordering::Relaxed);
-
-        if try_rpc_node_first {
-            let error = match self.get_full_checkpoint_from_primary(sequence_number).await {
-                Ok(checkpoint) => {
-                    self.reset_fullnode_failure_metrics();
-                    return Ok(checkpoint);
-                }
-                Err(error) => error,
-            };
 
-            self.num_failures.fetch_add(1, Ordering::Relaxed);
-            tracing::debug!(?error, "primary client error while fetching checkpoint");
-
-            let last_success = self.last_success.load(Ordering::Relaxed);
-            let num_failures = self.num_failures.load(Ordering::Relaxed);
-            if self.fallback_client.is_none()
-                || !self
-                    .fallback_client
-                    .as_ref()
-                    .unwrap()
-                    .is_eligible_for_fallback(sequence_number, &error, last_success, num_failures)
-            {
-                tracing::debug!(
-                    fallback_client_set = ?self.fallback_client.is_some(),
-                    "primary client error while fetching checkpoint is not eligible for fallback, \
-                    since last_success: {:?}, num_failures: {:?}",
-                    last_success.elapsed(),
-                    num_failures
-                );
-                return Err(error);
+        match self.download_handler.download_checkpoint_decision() {
+            CheckpointDownloadDecision::AttemptPrimary => {
+                self.handle_primary_attempt(sequence_number).await
+            }
+            CheckpointDownloadDecision::TryFallback(_) => {
+                self.handle_fallback_attempt(sequence_number).await
```

Comment:
Should we be logging the error in this case? (Currently it looks like we are ignoring it via `_` in the pattern.)